### PR TITLE
New endpoint for import/export dashboard with trimed json

### LINKED ISSUE #31044: A new endpoint for returning trimed dashboard json
A new endpoint similar to GET /api/dashboards/uid/:uid to get the trimmed dashboard json by using cue schema

### PR DESCRIPTION
**Pull request by ying-jeanne**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #31044

**Special notes for your reviewer**:

### CODE REVIEW

File: cue/data/gen.cue [hunk 1]
```diff
@@ -10,7 +10,7 @@ Family: scuemata.#Family & {
                 // TODO must isolate or remove identifiers local to a Grafana instance...?
                 id?: number
                 // Unique dashboard identifier that can be generated by anyone. string (8-40)
-                uid: string
+                uid?: string
```
Comment (ying-jeanne) [inline cue/data/gen.cue:13]: for post dashboard json, uid is not filled in the request, so the validation was failed. need uid to be optional

File: cue/data/gen.cue [hunk 2]
```diff
@@ -91,7 +92,7 @@ Family: scuemata.#Family & {
                     // _pv: { maj: int, min: int }
                     // The major and minor versions of the panel plugin for this schema.
                     // TODO 2-tuple list instead of struct?
-                    panelSchema: { maj: number, min: number }
+                    panelSchema?: { maj: number, min: number }
```
Comment (ying-jeanne) [inline cue/data/gen.cue:95]: for now, we don't have panelSchema yet, For the backward compatibility, i am wonder, should we put manually 0,0 after parsing the request.

File: pkg/api/dashboard.go [hunk 3]
```diff
@@ -46,6 +53,129 @@ func dashboardGuardianResponse(err error) response.Response {
 	return response.Error(403, "Access denied to this dashboard", nil)
 }
 
+func (hs *HTTPServer) GetTrimedDashboard(c *models.ReqContext) response.Response {
```
Comment (ying-jeanne) [inline pkg/api/dashboard.go:56]: planned to refactory, since a big part of it is reused in GetDashboard as well.

File: cue/data/gen.cue (hunk 2 above)
Comment (ryantxu) [reply cue/data/gen.cue:95]: This does not seem like something that should ever be optional -- can we hardcode 0,0 if not set?

Comment (ryantxu) [reply cue/data/gen.cue:95]: oh never mind -- this is on the dashboard, and yes will often be missing!

File: pkg/api/dashboard.go [hunk 4]
```diff
@@ -46,9 +46,29 @@ func dashboardGuardianResponse(err error) response.Response {
...
+	}
+
+	c.TimeRequest(metrics.MApiDashboardGet)
+	return response.JSON(200, dto)
+}
+
 func (hs *HTTPServer) GetDashboard(c *models.ReqContext) response.Response {
 	slug := c.Params(":slug")
 	uid := c.Params(":uid")
+	isTrimed := c.QueryBoolWithDefault("istrimed", false)
```
Comment (marefr) [inline pkg/api/dashboard.go:71]:
```suggestion
	trimDefaults := c.QueryBoolWithDefault("trimDefaults", false)
```

File: pkg/api/dashboard.go [hunk 5]
```diff
@@ -154,6 +174,14 @@ func (hs *HTTPServer) GetDashboard(c *models.ReqContext) response.Response {
 			return response.Error(500, "Error while loading library panels", err)
 		}
 	}
+	var trimedJson simplejson.Json
+	if isTrimed {
```
Comment (marefr) [inline pkg/api/dashboard.go:178]:
```suggestion
	if trimDefaults {
```

File: pkg/api/dashboard.go [hunk 6]
```diff
@@ -254,11 +282,17 @@ func (hs *HTTPServer) deleteDashboard(c *models.ReqContext) response.Response {
 }
 
 func (hs *HTTPServer) PostDashboard(c *models.ReqContext, cmd models.SaveDashboardCommand) response.Response {
+	var err error
 	cmd.OrgId = c.OrgId
 	cmd.UserId = c.UserId
-
+	isTrimed := c.QueryBoolWithDefault("istrimed", false)
```
Comment (marefr) [inline pkg/api/dashboard.go:288]:
```suggestion
	trimDefaults := c.QueryBoolWithDefault("trimDefaults", false)
```

File: pkg/api/dashboard.go [hunk 7]
```diff
@@ -254,11 +282,17 @@ func (hs *HTTPServer) deleteDashboard(c *models.ReqContext) response.Response {
...
+	if isTrimed {
```
Comment (marefr) [inline pkg/api/dashboard.go:289]:
```suggestion
	if trimDefaults {
```

File: pkg/api/dtos/dashboard.go [hunk 8]
```diff
@@ -37,6 +37,11 @@ type DashboardFullWithMeta struct {
 	Dashboard *simplejson.Json `json:"dashboard"`
 }
 
+type TrimDashboardFullWithMeta struct {
+	Meta      *simplejson.Json `json:"meta"`
```
Comment (marefr) [inline pkg/api/dtos/dashboard.go:41]: Not used currently, remove? Or maybe useful?

File: pkg/models/dashboards.go [hunk 9]
```diff
@@ -346,6 +346,12 @@ type SaveDashboardCommand struct {
 	Result *Dashboard
 }
 
+type TrimDashboardCommand struct {
+	Dashboard *simplejson.Json `json:"dashboard" binding:"Required"`
+	Meta      *simplejson.Json `json:"meta"`
```
Comment (marefr) [inline pkg/models/dashboards.go:351]: Not used currently, remove? Or maybe useful?

File: pkg/services/schemaloader/schemaloader.go [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+	"github.com/grafana/grafana/pkg/schema"
+	"github.com/grafana/grafana/pkg/schema/load"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/registry"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+func init() {
+	remotecache.Register(&RenderUser{})
```
Comment (marefr) [inline pkg/services/schemaloader/schemaloader.go:19]: What's this for?

**Review by marefr — COMMENTED** [review]
Added some comments/suggestions. 

As discussed, endpoint refactor looks good :+1: 
In regards to getting this merged asap I would suggest a feature toggle similar to alerting, ref https://github.com/grafana/grafana/blob/759a501564be89e0820c72fdb64e925fda606214/pkg/services/ngalert/ngalert.go#L107

File: pkg/models/dashboards.go (hunk 9 above)
Comment (ying-jeanne) [reply pkg/models/dashboards.go:351]: it is used by TrimDashboard, since meta is not built but directly take from input and returned in output

**Pull request closed without merging**